src/audio/copier/copier.c

```diff
@@ -780,6 +836,249 @@ static int copier_comp_trigger(struct comp_dev *dev, int cmd)
+			r_ptr + src_stream_sample_count : (int16_t *)src->end_addr;
+
+		r_ptr_before_loop = r_ptr;
+
+		do {
+			*w_ptr = *r_ptr;
+			r_ptr += src->channels;
+			w_ptr += dst->channels;
+		} while (r_ptr < r_end_ptr && w_ptr < (int16_t *)dst->end_addr);
+
+		src_stream_sample_count -= r_ptr - r_ptr_before_loop;
+	}
+}
+
+static void copy_single_channel_c32(struct audio_stream __sparse_cache *dst,
+				    int dst_channel,
+				    const struct audio_stream __sparse_cache *src,
+				    int src_channel, int frame_count)
+{
+	int32_t *r_ptr = (int32_t *)src->r_ptr + src_channel;
+	int32_t *w_ptr = (int32_t *)dst->w_ptr + dst_channel;
+
+	/* We have to iterate over frames here. However, tracking frames requires using
+	 * of expensive division operations (e.g., inside audio_stream_frames_without_wrap()).
+	 * So let's track samples instead. Since we only copy one channel, src_stream_sample_count
+	 * is NOT number of samples we need to copy but total samples for all channels. We just
+	 * track them to know when to stop.
+	 */
+	int src_stream_sample_count = frame_count * src->channels;
+
+	while (src_stream_sample_count) {
+		int src_samples_without_wrap;
+		int32_t *r_end_ptr, *r_ptr_before_loop;
+
+		r_ptr = audio_stream_wrap(src, r_ptr);
+		w_ptr = audio_stream_wrap(dst, w_ptr);
+
+		src_samples_without_wrap = audio_stream_samples_without_wrap_s32(src, r_ptr);
+		r_end_ptr = src_stream_sample_count < src_samples_without_wrap ?
+			r_ptr + src_stream_sample_count : (int32_t *)src->end_addr;
+
+		r_ptr_before_loop = r_ptr;
+
+		do {
+			*w_ptr = *r_ptr;
+			r_ptr += src->channels;
+			w_ptr += dst->channels;
+		} while (r_ptr < r_end_ptr && w_ptr < (int32_t *)dst->end_addr);
+
+		src_stream_sample_count -= r_ptr - r_ptr_before_loop;
+	}
+}
+
+static int demux_from_multi_endpoint_buffer(struct copier_data *cd)
+{
+	struct comp_buffer __sparse_cache *multi_buf_c, *endp_buf_c;
+	int endp_idx, endp_channel;
+	uint32_t avail_frames, avail_bytes, container_size;
+
+	multi_buf_c = buffer_acquire(cd->multi_endpoint_buffer);
+
+	/* NOTE: free space in destination buffers is not checked. This is done to have
+	 * behaviour consistent with copier_copy() which does not check for free space either.
+	 * However, quite probably, such behaviour is just a bug and have to be fixed in both
+	 * copier_copy() and here.
+	 */
+
+	avail_frames = audio_stream_get_avail_frames(&multi_buf_c->stream);
+	container_size = audio_stream_sample_bytes(&multi_buf_c->stream);
+
+	avail_bytes = avail_frames * audio_stream_frame_bytes(&multi_buf_c->stream);
+	buffer_stream_invalidate(multi_buf_c, avail_bytes);
+
+	for (endp_idx = 0; endp_idx < cd->endpoint_num; endp_idx++) {
+		uint32_t bytes_produced;
+
+		endp_buf_c = buffer_acquire(cd->endpoint_buffer[endp_idx]);
+
+		for (endp_channel = 0; endp_channel < endp_buf_c->stream.channels;
+				endp_channel++) {
+			int multi_buf_channel = endp_buf_c->chmap[endp_channel];
+
+			switch (container_size) {
```

Comment:
container_size is constant within both loops. Could be optimized, especially if this is called as part of pipeline processing.

Reply:
Modified to select channel copy func at copier_prepare().